File: src/main/java/selenium/elements/Radiobutton.java
package selenium.elements;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import selenium.Browser;

import java.security.InvalidParameterException;

public class Radiobutton {
    private By trueElementLocator;
    private By falseElementLocator;
    private By currentLocator;

    public Radiobutton(By elementLocator) {
        currentLocator = elementLocator;
    }

    public Radiobutton(By trueElementLocator, By falseElementLocator) {
        this.trueElementLocator = trueElementLocator;
        this.falseElementLocator = falseElementLocator;
    }

    public void click() {
        getLabel().click();
    }

    public void waitAndClick() {
        getLabel().waitAndClick();
    }

    public String getText() {
        return getLabel().getText();
    }

    public boolean isDisplayed() {
        return getLabel().isDisplayed();
    }

    public boolean isChecked() {
        return getElement().isSelected();
    }

    public boolean isUnchecked() {
        return !isChecked();
    }

    public WebElement getElement() {
        if (currentLocator == null && trueElementLocator != null) {
            throw new InvalidParameterException("Radiobutton cannot be called directly if constructed with true/false locators, call yes() or no() first");
        }
        return this.getDriver().findElement(currentLocator);
    }

    public Label getLabel() {
        return new Label(getElement());
    }

    private WebDriver getDriver() {
        return Browser.vanillaDriver();
    }

    public Radiobutton yes() {
        if (trueElementLocator == null) {
            throw new InvalidParameterException("Radiobutton must be constructed with true and false locators to use yes()-calls");
        }
        currentLocator = trueElementLocator;
        return this;
    }

    public Radiobutton no() {
        if (falseElementLocator == null) {
            throw new InvalidParameterException("Radiobutton must be constructed with true and false locators to use no()-calls");
        }
        currentLocator = falseElementLocator;
        return this;
    }
}

File: src/main/kotlin/no/sanchezrolfsen/framework/selenium/elements/Radiobutton.kt
package no.sanchezrolfsen.framework.selenium.elements

import no.sanchezrolfsen.framework.selenium.Browser
import org.openqa.selenium.By
import org.openqa.selenium.WebDriver
import org.openqa.selenium.WebElement
import java.security.InvalidParameterException

class Radiobutton {
    private var trueElementLocator: By? = null
    private var falseElementLocator: By? = null
    private var currentLocator: By? = null

    constructor(elementLocator: By) {
        currentLocator = elementLocator
    }

    constructor(trueElementLocator: By, falseElementLocator: By) {
        this.trueElementLocator = trueElementLocator
        this.falseElementLocator = falseElementLocator
    }

    fun click() {
        getLabel().click()
    }

    fun waitAndClick() {
        getLabel().waitAndClick()
    }

    fun getText(): String = getLabel().getText()

    fun isDisplayed(): Boolean = getLabel().isDisplayed()

    fun isChecked(): Boolean = getElement().isSelected

    fun isUnchecked(): Boolean = !isChecked()

    fun getElement(): WebElement {
        if (currentLocator == null && trueElementLocator != null) {
            throw InvalidParameterException("Radiobutton cannot be called directly if constructed with true/false locators, call yes() or no() first")
        }
        return getDriver().findElement(currentLocator!!)
    }

    fun getLabel(): Label = Label(getElement())

    private fun getDriver(): WebDriver = Browser.vanillaDriver()

    fun yes(): Radiobutton {
        val locator = trueElementLocator
            ?: throw InvalidParameterException("Radiobutton must be constructed with true and false locators to use yes()-calls")
        currentLocator = locator
        return this
    }

    fun no(): Radiobutton {
        val locator = falseElementLocator
            ?: throw InvalidParameterException("Radiobutton must be constructed with true and false locators to use no()-calls")
        currentLocator = locator
        return this
    }
}
